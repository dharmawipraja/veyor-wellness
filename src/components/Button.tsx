import { MouseEventHandler, useMemo } from "react";

type ButtonProps = {
  title: string;
  type?: "submit" | "reset" | "button";
  onClick?: MouseEventHandler<HTMLButtonElement>;
  variant?: 'PRIMARY' | 'SECONDARY';
}

const getButtonStyles = (variant: string) => {
  const styleMapping: Record<string, string> = {
    PRIMARY: "text-white bg-black ",
    SECONDARY: "w-32 p-2 text-black bg-white border border-gray-700"
  }

  return styleMapping[variant]
}

const Button = ({ title, type, onClick = () => {}, variant = 'PRIMARY' }: ButtonProps) => {
  const styles = getButtonStyles(variant);

  const renderButton = useMemo(() => (
    <button type={type} className={`${styles} max-w-prose p-2 rounded-md`} onClick={onClick}>
      <span>{title}</span>
    </button>
  ), [])

  return (
    renderButton
  )
}

export default Button