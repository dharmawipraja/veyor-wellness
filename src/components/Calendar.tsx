import { useMemo } from 'react'
import DatePicker, { ReactDatePickerCustomHeaderProps } from "react-datepicker";

import { useAppDispatch, useAppSelector } from '../hooks/useAppReducer';
import { AppDispatch } from '../store';
import { saveDate, saveTime } from '../store/appointmentSlice';
import { getAvailableSlot } from '../utils/appointment';

import "../css/datepicker.css";
import { addMonths, generateMonthYear, getMonthIndex, getMonthName } from '../utils/date';

const MAX_MONTH = 6;

const onChange = (dispatch: AppDispatch) => (date: Date | null) => {
  date && dispatch(saveDate(date.toISOString()));
};

const onSelectTime = (value: string, dispatch: AppDispatch) => () => {
  dispatch(saveTime(value))
}

const renderRadioButton = (value: string, selectedTime: string, dispatch: AppDispatch) => (
  <label key={value} className='flex gap-2'>
    <input type="radio" value={value} checked={value === selectedTime} onChange={onSelectTime(value, dispatch)} />
    {value}
  </label>
)

const renderCalendarHeader = ({
  date,
  changeYear,
  changeMonth,
  decreaseMonth,
  increaseMonth,
  prevMonthButtonDisabled,
  nextMonthButtonDisabled,
}: ReactDatePickerCustomHeaderProps) => {
  const months = generateMonthYear(MAX_MONTH);

  return (
    <div className='flex items-center justify-between h-10 m-3'>
      {<button className='px-5' onClick={decreaseMonth} disabled={prevMonthButtonDisabled}>
        {"<"}
      </button>}
      <select
        className='w-full h-full px-5 border border-gray-300 rounded-md'
        value={`${getMonthName(date)} ${date.getFullYear()}`}
        onChange={({ target: { value } }) => {
          const fullDate = value.split(' ');
          const month = getMonthIndex(fullDate[0])
          const year = Number(fullDate[1])

          changeMonth(month)
          changeYear(year)
        }}
      >
        {months.map((option) => (
          <option key={option} value={option}>
            {option}
          </option>
        ))}
      </select>

      <button className='px-5' onClick={increaseMonth} disabled={nextMonthButtonDisabled}>
        {">"}
      </button>
    </div>
  )
}

const Calendar = () => {
  const { date, time } = useAppSelector(state => state.appointment);
  const dispatch = useAppDispatch();
  const availableSlot = getAvailableSlot(date)
  const currentDate = new Date()

  const renderCalendar = useMemo(() => {
    return (
      <div>
        <DatePicker
          selected={new Date(date)}
          onChange={onChange(dispatch)}
          minDate={currentDate}
          maxDate={addMonths(currentDate, MAX_MONTH)}
          dateFormatCalendar={"MMMM yyyy"}
          inline
          renderCustomHeader={props => renderCalendarHeader(props)}
        />
      </div>
    )
  }, [date])

  const renderTime = useMemo(() => {
    return (
      <div className='flex flex-col items-start justify-start'>
        <p className='my-5 font-bold'>Please select a time</p>
        {availableSlot.map(value => renderRadioButton(value, time, dispatch))}
      </div>
    )
  }, [date, time])

  return (
    <div className='py-8'>
      {renderCalendar}
      {renderTime}
    </div>
  )
}

export default Calendar