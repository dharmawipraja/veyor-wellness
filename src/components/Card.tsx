import React, { MouseEventHandler, ReactNode } from "react";

type CardProps = {
  children: ReactNode;
  onClick?: MouseEventHandler<HTMLDivElement>
};

const Card: React.FC<CardProps> = ({ children, onClick }) => {
  return (
    <div onClick={onClick} className="flex flex-col items-start w-full p-5 border border-gray-300 rounded-md">
      {children}
    </div>
  )
}

export default Card