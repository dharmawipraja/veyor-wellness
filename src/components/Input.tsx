import React from 'react';
import { useController } from 'react-hook-form';

type InputProps = {
  control: any;
  name: string;
  rules?: any;
  label?: string;
  required?: boolean;
  placeholder?: string;
  style?: string;
}

const Input: React.FC<InputProps> = (props) => {
  const { field, fieldState } = useController(props)
  const { error } = fieldState;
  const { label, required, placeholder, style } = props;

  return (
    <div className={`flex flex-col items-start gap-1 p-1 ${error ? 'bg-red-200' : ''} ${style}`}>
      {label ? <span>{label} {required ? '*' : ''}</span> : <span className={`hidden ${error ? 'text-red-200' : 'text-white'} md:block`}>label</span>}
      <input className="w-full p-2 border border-gray-300 rounded-sm" type='text' placeholder={placeholder} {...field} />
      {error && <p className='text-sm text-red-700'>{error.message}</p>}
    </div>
  )
}

export default Input