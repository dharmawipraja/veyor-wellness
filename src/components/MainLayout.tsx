import { Outlet } from "react-router-dom"

const MainLayout = () => (
  <div>
    <header className="flex flex-col gap-10 mb-5">
      <h1 className="text-3xl">
        Book a wellness session.
      </h1>
      <p>Visit one of our expert consultants to get yourself feeling 100% again.</p>
    </header>
    <main>
      <Outlet />
    </main>
  </div>
)

export default MainLayout