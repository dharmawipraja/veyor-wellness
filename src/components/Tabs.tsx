import React, { ReactNode } from 'react';

type TabProps = {
  label: string;
  children: ReactNode;
};

type TabsProps = {
  activeTab: string;
  children: React.ReactElement<TabProps>[];
};

const Tabs: React.FC<TabsProps> = ({ children, activeTab }) => {
  return (
    <div className="mx-auto max-w-prose">
      <div className="tab">
        {children.map((child) => {
          const isActiveStyle =
            activeTab === child.props.label
              ? 'tab__item is-active'
              : 'tab__item';

          return (
            <div key={child.props.label} className={`sm:w-48 ${isActiveStyle}`}>
              {child.props.label}
            </div>
          );
        })}
      </div>
      <div className="py-4">
        {children.map((child) => {
          if (child.props.label === activeTab) {
            return <div key={child.props.label}>{child.props.children}</div>;
          }
          return null;
        })}
      </div>
    </div>
  );
};

const Tab: React.FC<TabProps> = ({ children }) => {
  return <>{children}</>;
};

export { Tabs, Tab };
