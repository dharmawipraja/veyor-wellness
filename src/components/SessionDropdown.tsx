import React, { useState } from "react";
import Card from "./Card";
import { SessionType } from "../types";
import { SESSIONS } from "../fixtures/appointmentData";

type SetSelectedData = React.Dispatch<React.SetStateAction<string>>;
type SetDropdown = React.Dispatch<React.SetStateAction<boolean>>;
type Props = {
  data: SessionType[]
  selectedData: string;
  setSelectedData: SetSelectedData
}

const openDropdown = (setIsDropdownOpen: SetDropdown) => () => {
  setIsDropdownOpen(true);
};

const onSelect = (
  item: string,
  setSelectedData: SetSelectedData,
  setIsDropdownOpen: SetDropdown
) => {
  return () => {
    setSelectedData(item);
    setIsDropdownOpen(false);
  };
};

const SessionDropdown: React.FC<Props> = ({ data, selectedData, setSelectedData }) => {
  const [isDropdownOpen, setIsDropdownOpen] = useState<boolean>(false);
  const label = SESSIONS.find(session => session.name === selectedData);

  return (
    <div>
      <Card onClick={openDropdown(setIsDropdownOpen)}>
        <p>{label?.name}</p>
        <span>{label?.duration} @ ${label?.price.toFixed(0)}</span>
      </Card>
      {isDropdownOpen && (
          <div className="absolute z-10 translate-y-6 bg-white border border-gray-300 rounded max-w-prose md:w-full">
            <ul className="flex flex-col items-start p-2">
              {data.map((item: SessionType) => (
                <li
                  key={item.name}
                  onClick={onSelect(item.name, setSelectedData, setIsDropdownOpen)}
                  className="flex flex-col items-start w-full gap-2 p-4 transition-all duration-200 rounded hover:bg-gray-200">
                  <p>{item.name}</p>
                  <span>{item.duration} @ ${item.price}</span>
                </li>
              ))}
            </ul>
          </div>
        )}
    </div>
  );
};

export default SessionDropdown;
