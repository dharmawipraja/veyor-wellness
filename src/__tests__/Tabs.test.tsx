import { render, screen } from '@testing-library/react';
import { Tabs, Tab } from '../components/Tabs';

describe('Tabs Component', () => {
  const renderTabs = (activeTab: string) => {
    render(
      <Tabs activeTab={activeTab}>
        <Tab label="Tab 1">Content for Tab 1</Tab>
        <Tab label="Tab 2">Content for Tab 2</Tab>
        <Tab label="Tab 3">Content for Tab 3</Tab>
      </Tabs>
    );
  };

  it('should match snapshot', () => {
    const { asFragment } = render(
      <Tabs activeTab={'Tab 2'}>
        <Tab label="Tab 1">Content for Tab 1</Tab>
        <Tab label="Tab 2">Content for Tab 2</Tab>
        <Tab label="Tab 3">Content for Tab 3</Tab>
      </Tabs>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders all tabs', () => {
    renderTabs('Tab 1');

    expect(screen.getByText('Tab 1')).toBeInTheDocument();
    expect(screen.getByText('Tab 2')).toBeInTheDocument();
    expect(screen.getByText('Tab 3')).toBeInTheDocument();
  });

  it('highlights the active tab', () => {
    renderTabs('Tab 2');

    const tab2 = screen.getByText('Tab 2');
    const tab1 = screen.getByText('Tab 1');
    const tab3 = screen.getByText('Tab 3');

    expect(tab1).toHaveClass('tab__item');
    expect(tab2).toHaveClass('tab__item is-active');
    expect(tab3).toHaveClass('tab__item');
  });

  it('displays the content of the active tab', () => {
    renderTabs('Tab 3');

    expect(screen.getByText('Content for Tab 3')).toBeInTheDocument();
    expect(screen.queryByText('Content for Tab 1')).not.toBeInTheDocument();
    expect(screen.queryByText('Content for Tab 2')).not.toBeInTheDocument();
  });

  it('does not display content for inactive tabs', () => {
    renderTabs('Tab 1');

    expect(screen.queryByText('Content for Tab 2')).not.toBeInTheDocument();
    expect(screen.queryByText('Content for Tab 3')).not.toBeInTheDocument();
  });
});
