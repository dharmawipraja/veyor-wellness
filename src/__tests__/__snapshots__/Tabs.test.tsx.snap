// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="mx-auto max-w-prose"
  >
    <div
      class="tab"
    >
      <div
        class="sm:w-48 tab__item"
      >
        Tab 1
      </div>
      <div
        class="sm:w-48 tab__item is-active"
      >
        Tab 2
      </div>
      <div
        class="sm:w-48 tab__item"
      >
        Tab 3
      </div>
    </div>
    <div
      class="py-4"
    >
      <div>
        Content for Tab 2
      </div>
    </div>
  </div>
</DocumentFragment>
`;
